raysect/core/math/function/float/function1d/interpolate.pyx

```diff
@@ -0,0 +1,538 @@
+
+    :param x: 1D memory view of the spline point x positions.
+    :param f: 1D memory view of the function value at spline point x positions.
+    """
+    def __init__(self, double[::1] x, double[::1] f):
+        self._x = x
+        self._f = f
+
+        cdef int n
+        n = len(x)
+        self._n = n
+        # Where 'a' has been calculated already the mask value = 1
+        self._mask_a = np.zeros((n - 1,), dtype=np.float64)
+        self._a = np.zeros((n - 1, 4), dtype=np.float64)
+        self._a_mv = self._a
+
+    @cython.initializedcheck(False)
+    @cython.boundscheck(False)
+    @cython.cdivision(True)
+    cdef double get_gradient(self, double[::1] x_spline, double[::1] y_spline, int index):
+        """
+        Calculate the normalised gradient at x_spline[index] based on the central difference approximation unless at
+        the edges of the array x_spline.
+
+        At x[i], the gradient is normally estimated using the central difference approximation [y[i-1], y[i+1]]/2
+        For a normalised range x[i], x[i+1] between 0 and 1, this is the same except for unevenly spaced data.
+        Unevenly spaced data has a normalisation x[i-1] - x[i+1] != 2, it is defined as x_eff in this function by
+        re-scaling the distance x[i-1] - x[i+1] using normalisation (x[i+1] - x[i]) = 1.
+
+        At the start and end of the array, the forward or backward difference approximation is calculated over
+        a  (x[i+1] - x[i]) = 1 or  (x[i] - x[i-1]) = 1 respectively. The end spline gradient is not used for
+        extrapolation
+
+        .. WARNING:: For speed, this function does not perform any zero division, type or bounds
+          checking. Supplying malformed data may result in data corruption or a
+          segmentation fault.
+
+        :param x_spline: A memory view to a double array containing monotonically increasing values.
+        :param y_spline: The desired spline points corresponding function returned values
+        :param int index: The index of the lower spline point that the gradient is to be calculated for
+        """
+        cdef double dfdx
+        cdef double x_eff
+        if index == 0:
+            dfdx = (y_spline[index + 1] - y_spline[index])
+        elif index == self._n - 1:
+            dfdx = y_spline[index] - y_spline[index - 1]
+        else:
+            # Finding the normalised distance x_eff
+            x_eff = (x_spline[index + 1] - x_spline[index - 1])/(x_spline[index + 1] - x_spline[index])
+            if x_eff != 0:
+                dfdx = (y_spline[index + 1]-y_spline[index - 1])/x_eff
+            else:
+                raise ZeroDivisionError('Two adjacent spline points have the same x value!')
+        return dfdx
+
+    @cython.cdivision(True)
+    @cython.boundscheck(False)
+    @cython.initializedcheck(False)
+    cdef double evaluate(self, double px, int index) except? -1e999:
+        cdef int index_use = find_index(self._x, px)
```

Comment:
The Interpolate1D evaluate method already obtained the relevant array index (passed in `index`), this does not need to be repeated (it is also really expensive!).